### PR TITLE
DEP: interpolate: deprecate interp2d

### PR DESCRIPTION
**Pull request by ev-br**
#### Reference issue

mailing list discussion: https://mail.python.org/archives/list/scipy-dev@python.org/thread/LU72CAGYFH37PZWCANMUF6UWFNSPZRHU/

#### What does this implement/fix?

Make `interp2d` raise DeprecationWarnings starting from scipy 1.10

#### Additional information

Posting early to increase exposure: the mailing list thread needs at least two weeks (i.e., this can go in not earlier than Oct 20).
Also the transition guide link is for now to a github gist, will likely need to be updated based on the final state of gh-17178.

### CODE REVIEW

**Comment by ev-br** [comment]
@j-bowhay Jake, would you mind pointing me to how to keep track of the deprecation? Especially given the three-step process, as suggested by Matthew in the mailing list?

**Comment by j-bowhay** [comment]
The status of all deprecation if currently traced in #15765 cc. @h-vetinari

File: scipy/interpolate/_interpolate.py [hunk 1]
```diff
@@ -322,6 +341,8 @@ def __call__(self, x, y, dx=0, dy=0, assume_sorted=False):
             z = z[0]
         return array(z)
 
+interp2d.__doc__ = interp2d.__doc__ % {"dep_mesg": dep_mesg}
```
Comment (j-bowhay) [inline scipy/interpolate/_interpolate.py:344]:
```suggestion

interp2d.__doc__ = interp2d.__doc__ % {"dep_mesg": dep_mesg}
```

File: scipy/interpolate/_interpolate.py [hunk 2]
```diff
@@ -92,11 +92,28 @@ def lagrange(x, w):
 # !! found, get rid of it!
 
 
+dep_mesg = """\
+`interp2d` is deprecated in SciPy 1.10 and will be removed in a future SciPy version.
```
Comment (j-bowhay) [inline scipy/interpolate/_interpolate.py:96]: I think it is better if we commit to a version. As a user I means I know when I have to change my code by and for maintainers it makes it more obvious when to execute the deprecation.
```suggestion
`interp2d` is deprecated in SciPy 1.10 and will be removed in SciPy 1.12.0. 
```

File: scipy/interpolate/_interpolate.py [hunk 3]
```diff
@@ -92,11 +92,28 @@ def lagrange(x, w):
...
+For more details see
+`https://gist.github.com/ev-br/8544371b40f414b7eaf3fe6217209bff`
+"""
+
 class interp2d:
     """
     interp2d(x, y, z, kind='linear', copy=True, bounds_error=False,
              fill_value=None)
 
+    %(dep_mesg)s
```
Comment (j-bowhay) [inline scipy/interpolate/_interpolate.py:115]:
```suggestion
    .. deprecated:: 1.10.0
        %(dep_mesg)s
```

**Review by j-bowhay — CHANGES_REQUESTED** [review]
A few comments to hopefully make this more consistent with other deprecations. Also I think it would be nice to have a test checking that the warning appears when it is expected to.

**Comment by ev-br** [comment]
OK, I think it's easiest to just copy-paste the message into the docstring. Will do together with eating-own-dogfood exercise, there is interp2d usage in scipy.stats to be ported over.

**Comment by ev-br** [comment]
Test failures are due to internal interp2d usage in scipy.stats, which gh-17198 removes.

**Comment by ev-br** [comment]
Rebased on main now that gh-17198 is in, and lumped together review suggestions from Jake.  

Apologies for the mass ping of *code owners*! This was a git mistake when pulling in changes made in the GH interface.

Recording the error here: apparently, when there is a review change via GH interface "suggested edit" and a local change, updating a PR needs *two* rebases: one `origin/<PR fork>` to pull in the reviewer edits, and the second on $upstream/main$ to sync the PR branch with the upstream main.

File: scipy/interpolate/_interpolate.py [hunk 4]
```diff
@@ -93,11 +93,42 @@ def lagrange(x, w):
...
+    For legacy code, nearly bug-for-bug compatible replacements are
+    `RectBivariateSpline` on regular grids, and `bisplrep`/`bisplev` for
+    scattered 2D data.
+
+    In new code, for regular grids use `RegularGridInterpolator` instead.
+    For scattered data, prefer `LinearNDInterpolator` or
+    `CloughTocher2DInterpolator`.
+
+    For more details see
+    `https://gist.github.com/ev-br/8544371b40f414b7eaf3fe6217209bff`
```
Comment (j-bowhay) [inline scipy/interpolate/_interpolate.py:129]:
```suggestion
       `interp2d` is deprecated in SciPy 1.10 and will be removed in SciPy
       1.12.0. 

       For legacy code, nearly bug-for-bug compatible replacements are
       `RectBivariateSpline` on regular grids, and `bisplrep`/`bisplev` for
       scattered 2D data.

       In new code, for regular grids use `RegularGridInterpolator` instead.
       For scattered data, prefer `LinearNDInterpolator` or
       `CloughTocher2DInterpolator`.

       For more details see
       `https://gist.github.com/ev-br/8544371b40f414b7eaf3fe6217209bff`
```

I think this needs to be indented to show up within the directive.
Before:
image
After:
image

File: scipy/interpolate/_interpolate.py [hunk 5]
```diff
@@ -93,11 +93,42 @@ def lagrange(x, w):
...
+`RectBivariateSpline` on regular grids, and `bisplrep`/`bisplev` for
+scattered 2D data.
+
+In new code, for regular grids use `RegularGridInterpolator` instead.
+For scattered data, prefer `LinearNDInterpolator` or
+`CloughTocher2DInterpolator`.
+
+For more details see
+`https://gist.github.com/ev-br/8544371b40f414b7eaf3fe6217209bff`
+"""
```
Comment (j-bowhay) [inline scipy/interpolate/_interpolate.py:109]:
```suggestion
dep_mesg = """\
'interp2d' is deprecated in SciPy 1.10 and will be removed in SciPy 1.12.0. 

For legacy code, nearly bug-for-bug compatible replacements are
'RectBivariateSpline' on regular grids, and 'bisplrep'/'bisplev' for
scattered 2D data.

In new code, for regular grids use 'RegularGridInterpolator' instead.
For scattered data, prefer 'LinearNDInterpolator' or
'CloughTocher2DInterpolator'.

For more details see
'https://gist.github.com/ev-br/8544371b40f414b7eaf3fe6217209bff'
"""
```
Minor enough to ignore if you want but in other deprecation reviews there is has been a preference for single quotes over backticks as they are easier to see in the terminal

File: scipy/interpolate/_interpolate.py (hunk 4 above)
Comment (ev-br) [reply scipy/interpolate/_interpolate.py:129]: Cool, I did not know about that!

File: scipy/interpolate/_interpolate.py (hunk 5 above)
Comment (ev-br) [reply scipy/interpolate/_interpolate.py:109]: Ugh. Neither passes the "speck on a monitor"  test for me, but if there's a preference, so be it

**Review by j-bowhay — APPROVED** [review]
Other than the lint errors this looks good to go now

**Comment by ev-br** [comment]
No objections surfaced in the ML discussion, so I'd like to merge this for 1.10 before branching. Am going to hit the merge button this week unless further comments. 

One  remaining glitch is that the porting recommendations are still in a gist under my GH account. This can be adjusted in a follow-up though, if we figure a better place for it.

**Comment by ev-br** [comment]
OK, the mailing list discussion outcome is positive, and there are no further comments either in the ML or here. Let's press on with this deprecation. Thanks Jake for the help, thanks to everybody who contributed to the ML discussion. Merging.

**Pull request merged**